vnext/Microsoft.ReactNative/JsiApi.cpp

```diff
@@ -58,137 +43,212 @@ struct JsiBufferWrapper : facebook::jsi::Buffer {
-}
+  static JsiObjectRef MakeJsiArrayData(facebook::jsi::Array &&arr) {
+    auto &&accessor = AsPointerAccessor(std::move(arr));
+    return {reinterpret_cast<uint64_t>(std::exchange(accessor.ptr_, nullptr))};
+  }
 
-static facebook::jsi::ArrayBuffer const &AsArrayBuffer(facebook::jsi::Runtime::PointerValue const **ptr) noexcept {
-  return *reinterpret_cast<facebook::jsi::ArrayBuffer const *>(ptr);
-}
+  static JsiWeakObjectRef MakeJsiWeakObjectData(facebook::jsi::WeakObject &&weakObj) {
+    auto &&accessor = AsPointerAccessor(std::move(weakObj));
+    return {reinterpret_cast<uint64_t>(std::exchange(accessor.ptr_, nullptr))};
+  }
 
-static facebook::jsi::Value const *AsValue(JsiValueRef const &data) noexcept {
-  return reinterpret_cast<facebook::jsi::Value const *>(&data);
-}
+  static JsiObjectRef MakeJsiFunctionData(facebook::jsi::Function &&func) {
+    auto &&accessor = AsPointerAccessor(std::move(func));
+    return {reinterpret_cast<uint64_t>(std::exchange(accessor.ptr_, nullptr))};
+  }
+};
 
-static JsiValueRef ToJsiValueData(facebook::jsi::Value const &value) noexcept {
-  return {static_cast<JsiValueKind>(static_cast<int32_t>(value.kind_)),
-          *reinterpret_cast<uint64_t const *>(&value.data_.number)};
-}
+struct ValueAccessor {
+  static JsiValueRef ToJsiValueData(facebook::jsi::Value const &value) noexcept {
+    ValueAccessor const &accessor = reinterpret_cast<ValueAccessor const &>(value);
+    return {static_cast<JsiValueKind>(static_cast<int32_t>(accessor.m_kind)),
+            *reinterpret_cast<uint64_t const *>(&accessor.m_data)};
+  }
 
-static JsiPropertyIdRef ToJsiPropertyNameIdData(facebook::jsi::PropNameID const &propNameId) noexcept {
-  return {reinterpret_cast<uint64_t>(propNameId.ptr_)};
-}
+  static JsiValueRef MakeJsiValueData(facebook::jsi::Value &&value) {
+    ValueAccessor &&accessor = reinterpret_cast<ValueAccessor &&>(value);
+    return {static_cast<JsiValueKind>(static_cast<int32_t>(std::exchange(accessor.m_kind, UndefinedKind))),
+            *reinterpret_cast<uint64_t *>(&accessor.m_data)};
+  }
 
-static JsiSymbolRef MakeJsiSymbolData(facebook::jsi::Runtime::PointerValue *pointerValue) {
-  return {reinterpret_cast<uint64_t>(pointerValue)};
-}
+ private:
+  enum ValueKind { UndefinedKind } m_kind;
+  double m_data;
 
-static JsiStringRef MakeJsiStringData(facebook::jsi::Runtime::PointerValue *pointerValue) {
-  return {reinterpret_cast<uint64_t>(pointerValue)};
-}
+  static_assert(sizeof(m_data) == sizeof(uint64_t), "Value data should fit in a 64-bit register");
+};
 
-static JsiObjectRef MakeJsiObjectData(facebook::jsi::Runtime::PointerValue *pointerValue) {
-  return {reinterpret_cast<uint64_t>(pointerValue)};
-}
+static_assert(
+    sizeof(ValueAccessor) == sizeof(facebook::jsi::Value),
+    "ValueAccessor must have the same size as facebook::jsi::Value");
+
+// RuntimeAccessor is used to access facebook::jsi::Runtime protected members.
+struct RuntimeAccessor : facebook::jsi::Runtime {
```

Comment:
final

Reply:
It must not be `final`. The class must call methods as virtual methods. If de-virtualization ever happened then we will have runtime crashes. The `final` is a very precise tool that must be used lightly.